File: DotNet/RestSample/WaveData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RestSample
{
    public class WaveData
    {
        public WaveData()
        {
        }

        public static WaveData Next(TimeSpan interval, double multiplier, int order)
        {
            DateTime now = DateTime.UtcNow;
            TimeSpan timeOfDay = now.TimeOfDay;

            double radians = ((timeOfDay.TotalSeconds % interval.TotalSeconds) / interval.TotalSeconds) * 2 * Math.PI;

            return new WaveData()
            {
                Order = order,
                Radians = radians,
                Tau = radians / (2 * Math.PI),
                Sin = multiplier * Math.Sin(radians),
                Cos = multiplier * Math.Cos(radians),
                Tan = multiplier * Math.Tan(radians),
                Sinh = multiplier * Math.Sinh(radians),
                Cosh = multiplier * Math.Cosh(radians),
                Tanh = multiplier * Math.Tanh(radians),
            };
        }

        public int Order
        {
            get;
            set;
        }

        public double Tau
        {
            get;
            set;
        }

        public double Radians
        {
            get;
            set;
        }

        public double Sin
        {
            get;
            set;
        }

        public double Cos
        {
            get;
            set;
        }

        public double Tan
        {
            get;
            set;
        }

        public double Sinh
        {
            get;
            set;
        }

        public double Cosh
        {
            get;
            set;
        }

        public double Tanh
        {
            get;
            set;
        }

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder();
            builder.AppendLine(String.Format("Order = {0}", Order));
            builder.AppendLine(String.Format("Radians = {0}", Radians));
            builder.AppendLine(String.Format("Tau     = {0}", Tau));
            builder.AppendLine(String.Format("Sine    = {0}", Sin));
            builder.AppendLine(String.Format("Cosine  = {0}", Cos));
            builder.AppendLine(String.Format("Tangent = {0}", Tan));
            builder.AppendLine(String.Format("Sinh    = {0}", Sinh));
            builder.AppendLine(String.Format("Cosh    = {0}", Cosh));
            builder.AppendLine(String.Format("Tanh    = {0}", Tanh));
            return builder.ToString();
        }

    }
}

File: Basic/JavaScript/NodeJs/WaveData.js
// WaveData.js
//
// Copyright (C) 2018 OSIsoft, LLC. All rights reserved.
//
// THIS SOFTWARE CONTAINS CONFIDENTIAL INFORMATION AND TRADE SECRETS OF
// OSIsoft, LLC.  USE, DISCLOSURE, OR REPRODUCTION IS PROHIBITED WITHOUT
// THE PRIOR EXPRESS WRITTEN PERMISSION OF OSIsoft, LLC.
//
// RESTRICTED RIGHTS LEGEND
// Use, duplication, or disclosure by the Government is subject to restrictions
// as set forth in subparagraph (c)(1)(ii) of the Rights in Technical Data and
// Computer Software clause at DFARS 252.227.7013
//
// OSIsoft, LLC
// 1600 Alvarado St, San Leandro, CA 94577

module.exports = {
    // Wave Data Object
    WaveData: function() {
        this.Order = null;
        this.Radians = null;
        this.Tau = null;
        this.Sin = null;
        this.Cos = null;
        this.Tan = null;
        this.Sinh = null;
        this.Cosh = null;
        this.Tanh = null;
    },

    // creates a new wave event
    NextWave: function(interval, multiplier, order) {
        now = new Date();
        midnight = new Date();
        midnight.setHours(0, 0, 0, 0);
        totalSecondsDay = (now -
            midnight);
        _interval = (interval - midnight);
        radians = ((totalSecondsDay % _interval) / _interval) * 2 * Math.PI;

        newWave = new this.WaveData();
        newWave.Order = order;
        newWave.Radians = radians;
        newWave.Tau = radians / (2 * Math.PI);
        newWave.Sin = multiplier * Math.sin(radians);
        newWave.Cos = multiplier * Math.cos(radians);
        newWave.Tan = multiplier * Math.tan(radians);
        newWave.Sinh = multiplier * Math.sinh(radians);
        newWave.Cosh = multiplier * Math.cosh(radians);
        newWave.Tanh = multiplier * Math.tanh(radians);

        return newWave;
    }
}
